src/tests/daos_racer.c

```diff
@@ -0,0 +1,473 @@
+	[ENUM_DKEY] = {.sub_test = enum_dkey},
+	[ENUM_AKEY] = {.sub_test = enum_akey},
+	[ENUM_REC] = {.sub_test = enum_rec},
+	[PUNCH_OBJ] = {.sub_test = punch_obj},
+	[PUNCH_DKEY] = {.sub_test = punch_dkey},
+	[PUNCH_AKEY] = {.sub_test = punch_akey},
+};
+
+void sub_tests_init(struct racer_sub_tests *tests, uint32_t bits)
+{
+	int i;
+	uint32_t mask;
+
+	for (i = 0; i < TEST_SIZE; i++) {
+		mask = 1 << i;
+		if (mask & bits)
+			tests[i].sub_test = racer_tests[i].sub_test;
+	}
+}
+
+static int
+racer_test_idx(struct racer_sub_tests *tests)
+{
+	int idx;
+
+	idx = ts_ctx.tsc_mpi_rank % TEST_SIZE;
+	while (tests[idx].sub_test == NULL)
+		idx = (idx + 1) % TEST_SIZE;
+
+	return idx;
+}
+
+static struct option ts_ops[] = {
+	{ "pool_uuid",	required_argument,	NULL,	'p' },
+	{ "cont_uuid",	required_argument,	NULL,	'c' },
+	{ "time",	required_argument,	NULL,	't' },
+	{ NULL,		0,			NULL,	0   },
+};
+
+int
+main(int argc, char **argv)
+{
+	struct timeval	tv;
+	daos_size_t	scm_size = (2ULL << 30); /* default pool SCM size */
+	daos_size_t	nvme_size = (8ULL << 30); /* default pool NVMe size */
+	d_rank_t	svc_rank  = 0;	/* pool service rank */
+	unsigned	duration = 60; /* seconds */
+	double		expire = 0;
+	int		idx;
+	bool		finish = false;
+	struct racer_sub_tests	sub_tests[TEST_SIZE] = { 0 };
+	int		rc;
+
+	MPI_Init(&argc, &argv);
+	MPI_Comm_rank(MPI_COMM_WORLD, &ts_ctx.tsc_mpi_rank);
+	MPI_Comm_size(MPI_COMM_WORLD, &ts_ctx.tsc_mpi_size);
+	while ((rc = getopt_long(argc, argv,
+				 "p:c:t:",
+				 ts_ops, NULL)) != -1) {
+		char	*endp;
+
+		switch (rc) {
+		default:
+			fprintf(stderr, "Unknown option %c\n", rc);
+			return -1;
+		case 'p':
+			rc = uuid_parse(optarg, ts_ctx.tsc_pool_uuid);
+			if (rc)
+				return rc;
+			break;
+		case 'c':
+			rc = uuid_parse(optarg, ts_ctx.tsc_cont_uuid);
+			if (rc)
+				return rc;
+		case 't':
+			duration = strtoul(optarg, &endp, 0);
+			break;
+		}
+	}
+
+	if (seed == 0) {
+		gettimeofday(&tv, NULL);
+		seed = tv.tv_usec;
+	}
+	srand(seed);
+
+	if (ts_ctx.tsc_mpi_rank == 0 && uuid_is_null(ts_ctx.tsc_pool_uuid))
+		uuid_generate(ts_ctx.tsc_pool_uuid);
+	if (ts_ctx.tsc_mpi_rank == 0 && uuid_is_null(ts_ctx.tsc_cont_uuid))
+		uuid_generate(ts_ctx.tsc_cont_uuid);
+
+	ts_ctx.tsc_svc.rl_nr = 1;
+	ts_ctx.tsc_svc.rl_ranks  = &svc_rank;
+	ts_ctx.tsc_scm_size	= scm_size;
+	ts_ctx.tsc_nvme_size	= nvme_size;
+
+	if (ts_ctx.tsc_mpi_rank == 0) {
+		fprintf(stdout,
+			"racer start with %d threads duration %u secs\n"
+			"\tpool size     : SCM: %u MB, NVMe: %u MB\n",
+			ts_ctx.tsc_mpi_size, duration,
+			(unsigned int)(scm_size >> 20),
+			(unsigned int)(nvme_size >> 20));
+	}
+
+	rc = dts_ctx_init(&ts_ctx);
+	if (rc)
+		D_GOTO(out, rc);
+
+	sub_tests_init(sub_tests, 0xFFFF);
+	if (ts_ctx.tsc_mpi_rank == 0 && duration != 0)
+		expire = dts_time_now() + duration;
+
+	idx = racer_test_idx(sub_tests);
+	MPI_Barrier(MPI_COMM_WORLD);
+	while (1) {
+		sub_tests[idx].sub_test();
+		if (ts_ctx.tsc_mpi_rank == 0) {
```

Comment:
well, we need to check for expire on all the ranks otherwise all the other ranks will continue forever.

Reply:
hmm, rank 0 will check for expire and set finish, then all other ranks will check finish in the while loop.

Reply:
this would work in shared memory program, but this is MPI (different processes). you need to set expire and check it on all  ranks, or you have to do a non-blocking bcast before the start from all ranks except rank 0 and test it on every iteration, and when rank 0 checks expire and sets finish, you do the bcast from rank 0 to all other ranks.
probably the first solution is just easier since it seems the processes operate independently..

Reply:
Ah, you are right, forget this is MPI on different node. I am dumb here.